api/lib/infrastructure/repositories/answer-repository.js

```diff
@@ -1,114 +1,134 @@
+const { NotFoundError } = require('../../domain/errors');
 const Answer = require('../../domain/models/Answer');
 const answerStatusDatabaseAdapter = require('../adapters/answer-status-database-adapter');
-const BookshelfAnswer = require('../orm-models/Answer');
-const BookshelfKnowledgeElement = require('../orm-models/KnowledgeElement');
 
-const Bookshelf = require('../bookshelf');
-const { NotFoundError } = require('../../domain/errors');
-const jsYaml = require('js-yaml');
-const _ = require('lodash');
-
-function _adaptModelToDb(answer) {
+function _adaptAnswerToDb(answer) {
   return {
-    id: answer.id,
+    ..._.pick(answer, ['value', 'timeout', 'challengeId', 'assessmentId', 'timeSpent']),
     result: answerStatusDatabaseAdapter.toSQLString(answer.result),
     resultDetails: jsYaml.dump(answer.resultDetails),
-    value: answer.value,
-    timeout: answer.timeout,
-    challengeId: answer.challengeId,
-    assessmentId: answer.assessmentId,
-    timeSpent: answer.timeSpent,
   };
 }
 
-function _toDomain(bookshelfAnswer) {
-  if (bookshelfAnswer) {
-    return new Answer({
-      id: bookshelfAnswer.get('id'),
-      result: answerStatusDatabaseAdapter.fromSQLString(bookshelfAnswer.get('result')),
-      resultDetails: bookshelfAnswer.get('resultDetails'),
-      timeout: bookshelfAnswer.get('timeout'),
-      value: bookshelfAnswer.get('value'),
-      assessmentId: bookshelfAnswer.get('assessmentId'),
-      challengeId: bookshelfAnswer.get('challengeId'),
-      timeSpent: bookshelfAnswer.get('timeSpent'),
-    });
-  }
-  return null;
+function _adaptKnowledgeElementToDb(knowledgeElement) {
+  return _.pick(knowledgeElement, ['source', 'status', 'earnedPix',
+    'answerId', 'assessmentId', 'skillId', 'userId', 'competenceId']);
+}
+
+function _toDomain(answerDTO) {
+  return new Answer({
+    ...answerDTO,
+    result: answerStatusDatabaseAdapter.fromSQLString(answerDTO.result),
+  });
 }
 
+const FIELDS = Object.freeze(['id', 'result', 'resultDetails', 'timeout', 'value', 'assessmentId', 'challengeId', 'timeSpent']);
+
 module.exports = {
 
-  get(answerId) {
-    return BookshelfAnswer.where('id', answerId)
-      .fetch()
-      .then(_toDomain)
-      .catch((error) => {
-        if (error instanceof BookshelfAnswer.NotFoundError) {
-          throw new NotFoundError(`Not found answer for ID ${answerId}`);
-        }
-
-        throw error;
-      });
+  async get(id) {
+    const answerDTO = await knex
+      .select(FIELDS)
+      .from('answers')
+      .where({ id })
+      .first();
+
+    if (!answerDTO) {
+      throw new NotFoundError(`Not found answer for ID ${id}`);
+    }
+
+    return _toDomain(answerDTO);
   },
 
-  findByIds(answerIds) {
-    return BookshelfAnswer.where('id', 'in', answerIds)
-      .fetchAll()
-      .then((answers) => answers.models.map(_toDomain));
+  async findByIds(ids) {
+    const answerDTOs = await knex
+      .select(FIELDS)
+      .from('answers')
+      .whereIn('id', ids)
```

Comment:
*suggestion*
```suggestion
      .where({ id: knex.raw('any(?)', [ids]) })
```

Avec `whereIn`, `knex` génère malheureusement des requêtes SQL à rallonge du type `SELECT * FROM answers WHERE id IN ($1, $2, $3, …)` qui en plus d'être lourdes à soumettre vont être différentes à chaque appel (en fonction du nombre d'`id`) et compliquer le suivi des statistiques par requête (puisque le nombre de requêtes distinctes monte en flèche).

Le passage d'un unique paramètre de type tableau contourne élégamment le problème. Je pense qu'on devrait appliquer le même pattern aux autres `whereIn`, peut-être que ça aurait du sens de factoriser la partie `knex.raw('any(?)', [ … ]`) dans un _helper_.

Reply:
Accessoirement, il me semble que cette méthode `findByIds` n'est plus jamais appelée…

Reply:
Non mais attends, ça fait pas moins de deux méthodes de répo qui n'existent plus ! Un petit audit pourrait être nécessaire (et une plus grande attention aux détails dans les revues). Merci.
Et oui pour le IN -> ANY, c'est très clair

Reply:
j'ai ajouté un commit dans ce sens, j'attends les retours